okio/src/main/java/okio/RealBufferedSource.java

```diff
@@ -442,4 +471,54 @@ public boolean rangeEquals(long offset, ByteString bytes, int bytesOffset, int b
   @Override public String toString() {
     return "buffer(" + source + ")";
   }
+
+  final class BomAwareReader extends Reader {
+    final Charset suppliedCharset;
+    final ByteOrderMark[] byteOrderMarks;
+    private Reader delegate;
+
+    BomAwareReader(Charset charset, ByteOrderMark... byteOrderMarks) {
+      this.suppliedCharset = charset;
+      this.byteOrderMarks = byteOrderMarks;
+    }
+
+    @Override public int read(char[] data, int offset, int charCount) throws IOException {
+      if (delegate == null) {
+        if (closed) throw new IOException("closed");
+
+        // Make sure buffer holds all bytes of the BOM.
+        if (buffer.size() < ByteOrderMark.MAX_BOM_LENGTH) {
```

Comment:
Sadly this is not at all safe. Imagine a socket that returns `AB` and then stalls waiting for more data to be generated. In this code the user doesn’t see `A` or `B` until at least 4 characters show up. That’s a big bug.

(This happens for real if the document is JSON and the content is `{}`.)